File: apps/php-sdk/src/Models/BatchScrapeOptions.php
<?php

declare(strict_types=1);

namespace Firecrawl\Models;

final class BatchScrapeOptions
{
    private function __construct(
        private readonly ?ScrapeOptions $options = null,
        private readonly string|WebhookConfig|null $webhook = null,
        private readonly ?string $appendToId = null,
        private readonly ?bool $ignoreInvalidURLs = null,
        private readonly ?int $maxConcurrency = null,
        private readonly ?bool $zeroDataRetention = null,
        private readonly ?string $idempotencyKey = null,
        private readonly ?string $integration = null,
    ) {}

    public static function with(
        ?ScrapeOptions $options = null,
        string|WebhookConfig|null $webhook = null,
        ?string $appendToId = null,
        ?bool $ignoreInvalidURLs = null,
        ?int $maxConcurrency = null,
        ?bool $zeroDataRetention = null,
        ?string $idempotencyKey = null,
        ?string $integration = null,
    ): self {
        return new self(
            $options, $webhook, $appendToId, $ignoreInvalidURLs,
            $maxConcurrency, $zeroDataRetention, $idempotencyKey, $integration,
        );
    }

    /** @return array<string, mixed> */
    public function toArray(): array
    {
        $fields = [
            'webhook' => $this->webhook instanceof WebhookConfig ? $this->webhook->toArray() : $this->webhook,
            'appendToId' => $this->appendToId,
            'ignoreInvalidURLs' => $this->ignoreInvalidURLs,
            'maxConcurrency' => $this->maxConcurrency,
            'zeroDataRetention' => $this->zeroDataRetention,
            'integration' => $this->integration,
        ];

        $data = array_filter($fields, fn (mixed $v): bool => $v !== null);

        // Flatten scrape options into body (API expects top-level, not nested)
        if ($this->options !== null) {
            $data = array_merge($this->options->toArray(), $data);
        }

        return $data;
    }

    public function getIdempotencyKey(): ?string
    {
        return $this->idempotencyKey;
    }

    public function getOptions(): ?ScrapeOptions
    {
        return $this->options;
    }
}

File: apps/ruby-sdk/lib/firecrawl/models/batch_scrape_options.rb
# frozen_string_literal: true

module Firecrawl
  module Models
    # Options for a batch scrape job.
    class BatchScrapeOptions
      FIELDS = %i[
        options webhook append_to_id ignore_invalid_urls
        max_concurrency zero_data_retention idempotency_key integration
      ].freeze

      attr_reader(*FIELDS)

      def initialize(**kwargs)
        FIELDS.each { |f| instance_variable_set(:"@#{f}", kwargs[f]) }
      end

      def to_h
        {
          "options" => options&.to_h,
          "webhook" => webhook.is_a?(String) ? webhook : webhook&.to_h,
          "appendToId" => append_to_id,
          "ignoreInvalidURLs" => ignore_invalid_urls,
          "maxConcurrency" => max_concurrency,
          "zeroDataRetention" => zero_data_retention,
          "integration" => integration,
        }.compact
      end
    end
  end
end
